### ISSUE TITLE
revel run, module errors in freshly build $GOPATH with latest go binary

### ISSUE THREAD

**Issue by magoo** [revel/revel#1400]
TL;DR, receiving various panic and errors with `revel/cmd` in new environments with different go versions.

I have a _functional_ revel codebase w/ the following environment:

```
Revel Framework /Users/magoo/gopath/src/www-forecast/vendor/github.com/revel/revel /Users/magoo/gopath/src
Revel Version = "0.20.0"
Revel BuildDate = "2018-09-23"
Revel MinimumGoVersion = ">= go1.8"

Revel Command Utility Tool
Version 0.20.2
Build Date 2018-10-02
Minimum Go Version >= go1.8
Compiled By   go1.9.4 darwin/amd64

go version go1.11.1 darwin/amd64
```

People running my codebase in newly built $GOPATHs and go binaries are reporting the following, which I can reproduce. I have the following _not functional environment_ that this codebase is not functional.

```
Revel Framework /Users/magoo/testgopath/src/www-forecast/vendor/github.com/revel/revel /Users/magoo/testgopath/src
Revel Version = "0.20.0"
Revel BuildDate = "2018-09-23"
Revel MinimumGoVersion = ">= go1.8"

Revel Command Utility Tool
Version 0.20.2
Build Date 2018-10-02
Minimum Go Version >= go1.8
Compiled By   go1.11.1 darwin/amd64

go version go1.11.1 darwin/amd64
```

First, I clone my revel codebase.

Then, I get revel cmd: 

`go get -u github.com/revel/cmd/revel`

Then I `revel run` and get the following error:

```
WARN  11:10:12 source_info.go:100: Type found in package: controllers, but did not embed from: revel.Controller name=JSONResponse importpath=www-forecast/app/controllers foundstructures=Auth,Estimate,Scenario,Rank,Home,Forecast,Range,Sort, 
ERROR 11:10:15  revel module.go:163: Failed to load module.  Import of path failed  section=module modulePath=github.com/revel/modules/static error="cannot find package \"github.com/revel/modules/static\" in any of:\n\t/Users/magoo/testgopath/src/www-forecast/vendor/github.com/revel/modules/static (vendor tree)\n\t/usr/local/go/src/github.com/revel/modules/static (from $GOROOT)\n\t/Users/magoo/testgopath/src/github.com/revel/modules/static (from $GOPATH)" 
INFO  11:10:15    app     run.go:29: Running revel server                      
CRIT  11:10:15  revel logger.go:195: NewRoute: Failed to find controller for route path action /public/*filepath?Static.Serve 
...
```

I notice at this point that my `static` module directory is also missing(?). I can observe it reliably being deleted after `revel run`, which seems problematic.

This seems related to dep and modules being discussed in a few places, so I try: 

`go get -u github.com/revel/modules`
`go get -u github.com/revel/revel`

I try `revel run` again. It removes the error but still panics.

```
CRIT  11:22:52  revel logger.go:195: NewRoute: Failed to find controller for route path action /public/*filepath?Static.Serve 
```

I notice in my good environment, I observe this log:

```
WARN  10:52:51 source_info.go:100: Type found in package: controllers, but did not embed from: revel.Controller name=JSONResponse importpath=www-forecast/app/controllers foundstructures=Home,Forecast,Auth,Scenario,Range,Sort,Estimate,Rank,Static,TestRunner, 
```

and in the bad environment, Static and TestRunner modules do not appear before the panic:

```
WARN  11:22:50 source_info.go:100: Type found in package: controllers, but did not embed from: revel.Controller name=JSONResponse importpath=www-forecast/app/controllers foundstructures=Range,Sort,Estimate,Rank,Home,Forecast,Auth,Scenario, 
```

**Comment by notzippy** [comment]
It appears you are using a `vendor` folder for your project `/Users/magoo/testgopath/src/www-forecast/vendor/github.com/revel/modules/static`, this will lock in a specific version of Revel.  
Can you post the contents of your `Gopkg.toml` ?

**Comment by magoo** [comment]
Sure!

```
[[constraint]]
  name = "github.com/aws/aws-sdk-go"
  version = "1.12.79"

[[constraint]]
  name = "github.com/google/uuid"
  version = "0.2.0"

[[constraint]]
  name = "github.com/revel/modules"
  version = "0.20.0"

[[constraint]]
  branch = "master"
  name = "google.golang.org/api"

[prune]
  go-tests = true
  unused-packages = true
```

**Comment by notzippy** [comment]
It looks like you are missing revel/revel ?
```
[[constraint]]
  name = "github.com/revel/revel"
  version = "0.20.0"
```

**Comment by notzippy** [comment]
If you include modules and not revel in the vendor what happens is that the modules controllers inherit from a different revel.controller then your app does revel.controller which means when the tool looks to build the code the modules controllers do not inherit from revel.controller (even though they do) Go's package management system is finicky.

**Comment by magoo** [comment]
Just added that, same error.

I do suspect vendoring is where my problem lies. I just noticed that `revel run` invokes `dep` which I did not expect. It still does not find `TestRunner` and `Static` like my functioning build environment.

```
Revel executing: run a Revel application
WARN  12:57:44 source_info.go:102: Type found in package: controllers, but did not embed from: revel.Controller name=JSONResponse importpath=www-forecast/app/controllers foundstructures=Home,Forecast,Auth,Range,Sort,Estimate,Scenario,Rank, 
ERROR 12:57:50  revel module.go:163: Failed to load module.  Import of path failed  section=module modulePath=github.com/revel/modules/static error="cannot find package \"github.com/revel/modules/static\" in any of:\n\t/Users/magoo/testgopath/src/www-forecast/vendor/github.com/revel/modules/static (vendor tree)\n\t/usr/local/go/src/github.com/revel/modules/static (from $GOROOT)\n\t/Users/magoo/testgopath/src/github.com/revel/modules/static (from $GOPATH)" 
INFO  12:57:50    app     run.go:29: Running revel server   
```

**Comment by notzippy** [comment]
does this path exist ? `/Users/magoo/testgopath/src/www-forecast/vendor/github.com/revel/modules/static ` it should because it is in your Gopkg.toml file but it appears like it doesnt

**Comment by magoo** [comment]
Hmm. That path is not in my `Gopkg.toml` that I can see? How are you reasoning that it is there?

To answer: That path _does_ exist in my "good" environment. It does _not_ exist in my "bad" environment. In fact: it gets deleted from the cloned repository after running `revel run` in the "bad" environment. Is that a result of `dep` managing the vendor folder? 

1. First I see modules/testrunner deleted
2. Then, modules/static is deleted
3. Then, modules/testrunner is created, with only testrunner.go.

I can observe this by watching the modules folder in the osx Finder as `revel run` executes.

**Comment by notzippy** [comment]
That is it you need 
```
[[constraint]]
  name = "github.com/revel/modules"
  version = "0.20.1"
```
version `0.20.0` was missing the `module.go` file in the root of `github.com/revel/modules` which messes with deps. Try updating the version

**Comment by magoo** [comment]
I've also tried pinning the version go Gopkg.toml with `version = "=0.20.1"`, yet it is still pulling in what looks like `0.20.0`. I realize this is looking like a dep behavior but would love any insight to confirm what is going wrong.

**Comment by notzippy** [comment]
That seems really weird - here is my toml file
```
required = ["github.com/revel/revel", "github.com/revel/modules"]

# Note to use a specific version changes this to
[[override]]
  version = "0.20.1"
  name = "github.com/revel/modules"

# Note to use a specific version changes this to
#
[[override]]
  version = "0.20.0"
  name = "github.com/revel/revel"

[[override]]
  branch = "master"
  name = "github.com/revel/log15"

[[override]]
  branch = "master"
  name = "github.com/revel/cron"

[[override]]
  branch = "master"
  name = "github.com/xeonx/timeago"


```

and this is the result of run dep on an empty vendor folder
```
$ rm -rf vendor/*
$ dep ensure
$ ls vendor/github.com/revel/modules/
auth  csrf  db  jobs  LICENSE  modules.go  orm  pprof  README.md  server-engine  static  template-engine  testrunner
```

**Comment by magoo** [comment]
OK progress. That finally brought `modules.go` into vendor, but brings about another error I can't quite decipher.

```
Error stack [utils.go:275 command_config.go:195 revel_container.go:90 revel_container.go:222 revel_container.go:182 run.go:123 revel.go:103 proc.go:195 asm_amd64.s:2337]
ERROR 16:42:40 command_config.go:196: Failed to import package                 error="exit status 1" gopath=/Users/magoo/gopath GO-ROOT=/usr/local/Cellar/go/1.9.4/libexec output="Fetching sources...\n\nFailed to add the dependencies:\n\n  ✗ nothing to -add, github.com/revel/modules is already in Gopkg.toml and the project's direct imports or required list\n\nadding dependencies failed\n" 
ERROR 16:42:40  revel.go:104: Unable to execute                        error="Revel paths[error Failed to load module.  Import of path failed modulePath:github.com/revel/modules/static error:Module github.com/revel/modules/static was found outside of path /Users/magoo/gopath/src/github.com/revel/modules/static. ]" 

For reference, my .toml is updated per your feedback to:

```
required = ["github.com/revel/revel", "github.com/revel/modules"]

[[override]]
  version = "0.20.1"
  name = "github.com/revel/modules"

[[override]]
  name = "github.com/revel/revel"
  version = "0.20.0"

[[constraint]]
  name = "github.com/aws/aws-sdk-go"
  version = "1.12.79"

[[constraint]]
  name = "github.com/google/uuid"
  version = "0.2.0"

[[constraint]]
  branch = "master"
  name = "google.golang.org/api"

[prune]
  go-tests = true
  unused-packages = true
```

**Comment by magoo** [comment]
Yeah I am not getting all the modules loaded like you are. Something is keeping state that I am not familiar with.

**Comment by notzippy** [comment]
I think you need to remove those two packages from your gopath

`rm -rf  /Users/magoo/gopath/src/github.com/revel/modules/  /Users/magoo/gopath/src/github.com/revel/revel`

and the compiled folder

`rm -rf  /Users/magoo/gopath/pkg/{{architecture}}/github.com/revel/modules/  /Users/magoo/gopath/pkg/{{architecture}}/github.com/revel/revel`

I can tell more if you run the `revel` command with the `-v` option - like `revel run -a yourapp -v`

**Comment by magoo** [comment]
Thanks for your patience on this. Deleted those folders, same error. 

From the app directory: 

`revel run -v`

```
INFO  21:28:48 command_config.go:124: Checking import path /Users/magoo/gopath/src/www-forecast with /Users/magoo/gopath 
INFO  21:28:48 command_config.go:133: Updated import path                      path=www-forecast 
INFO  21:28:48 command_config.go:140: Returned import path                     path=www-forecast buildpath=/Users/magoo/gopath 
Revel executing: run a Revel application
INFO  21:28:48 command_config.go:243: Source root                              path=/Users/magoo/gopath cwd=/Users/magoo/gopath/src/www-forecast gopath=/Users/magoo/gopath 
INFO  21:28:48 command_config.go:257: Set application path                     path=/Users/magoo/gopath/src/www-forecast 
INFO  21:28:48 command_config.go:150: InitPackageResolver                      useVendor=true path=/Users/magoo/gopath/src/www-forecast 
INFO  21:28:48 command_config.go:158: Vendor folder detected, scanning for deps in path 
INFO  21:28:48   file.go:346: Seeking app package                      app=www-forecast 
INFO  21:28:48   file.go:354: Seeking remote package                   using=www-forecast remote=github.com/revel/revel 
INFO  21:28:48 github.com/revel/modules/static revel_container.go:221: Missing module                            error="cannot find package \"github.com/revel/modules/static\" in any of:\n\t/Users/magoo/gopath/src/www-forecast/vendor/github.com/revel/modules/static (vendor tree)\n\t/usr/local/Cellar/go/1.9.4/libexec/src/github.com/revel/modules/static (from $GOROOT)\n\t/Users/magoo/gopath/src/github.com/revel/modules/static (from $GOPATH)" 
INFO  21:28:48 command_config.go:173: Request for package                      package=github.com/revel/modules/static use vendor=true 
INFO  21:28:48 command_config.go:175: Using dependency manager to import package package=github.com/revel/modules/static 
INFO  21:28:48 command_config.go:192: Go get command                           exec=/usr/local/bin/dep dir=/Users/magoo/gopath/src/www-forecast args="[/usr/local/bin/dep ensure -add github.com/revel/modules/static]"

(env removed)

package=github.com/revel/modules/static 
Error stack [utils.go:275 command_config.go:195 revel_container.go:90 revel_container.go:222 revel_container.go:182 run.go:123 revel.go:103 proc.go:195 asm_amd64.s:2337]
ERROR 21:28:48 command_config.go:196: Failed to import package                 error="exit status 1" gopath=/Users/magoo/gopath GO-ROOT=/usr/local/Cellar/go/1.9.4/libexec output="Fetching sources...\n\nFailed to add the dependencies:\n\n  ✗ nothing to -add, github.com/revel/modules is already in Gopkg.toml and the project's direct imports or required list\n\nadding dependencies failed\n" 
INFO  21:28:48 build_error.go:18: Revel paths                              error="Failed to load module.  Import of path failed modulePath:github.com/revel/modules/static error:cannot find package \"github.com/revel/modules/static\" in any of:\n\t/Users/magoo/gopath/src/www-forecast/vendor/github.com/revel/modules/static (vendor tree)\n\t/usr/local/Cellar/go/1.9.4/libexec/src/github.com/revel/modules/static (from $GOROOT)\n\t/Users/magoo/gopath/src/github.com/revel/modules/static (from $GOPATH) " 
INFO  21:28:48 build_error.go:23: Stack                                    stack="[github.com/revel/cmd/logger/utils.go:275 github.com/revel/cmd/utils/build_error.go:22 github.com/revel/cmd/utils/build_error.go:36 revel/run.go:125 revel/revel.go:103 runtime/proc.go:195 runtime/asm_amd64.s:2337]" 
ERROR 21:28:48  revel.go:104: Unable to execute                        error="Revel paths[error Failed to load module.  Import of path failed modulePath:github.com/revel/modules/static error:cannot find package \"github.com/revel/modules/static\" in any of:\n\t/Users/magoo/gopath/src/www-forecast/vendor/github.com/revel/modules/static (vendor tree)\n\t/usr/local/Cellar/go/1.9.4/libexec/src/github.com/revel/modules/static (from $GOROOT)\n\t/Users/magoo/gopath/src/github.com/revel/modules/static (from $GOPATH) ]" 

```

**Comment by notzippy** [comment]
Does this path exist `/Users/magoo/gopath/src/www-forecast/vendor/github.com/revel/modules/static` ?

**Comment by notzippy** [comment]
Also sent you a message on gitter.im

**Comment by magoo** [comment]
Yes that path exists.

```
➜  static git:(master) ✗ pwd
/Users/magoo/gopath/src/www-forecast/vendor/github.com/revel/modules/static
➜  static git:(master) ✗ tree
.
└── app
    ├── controllers
    │   └── static.go
    └── model
        └── file_information.go

3 directories, 2 files

```

**Comment by magoo** [comment]
`github.com/revel/modules/static` exists in my vendor path but not my $GOPATH.

**Comment by notzippy** [comment]
The issue was the section of the Gopkg file, with `unused-packages = true` dep removed important packages that the Dep tool did not realize were needed. Removing that section fixed the issue

```
[prune]
  go-tests = true
#  unused-packages = true
```
